File: KimchilyWebGL/KimchilyCreator/Assets/World/Character.ts
import { KimchilyScriptBehaviour } from 'Kimchily.Script';
import { Debug, GameObject, Time, Vector3, WaitForSeconds } from 'UnityEngine';
import { sampleHeight } from './Motion';

export default class Character extends KimchilyScriptBehaviour {
    public beacon!: GameObject;
    public speed: number = 140;
    public blinkSeconds: number = 0.5;

    private elapsed: number = 0;
    private counters: Map<string, number> = new Map<string, number>();

    Start(): void {
        this.counters.set('starts', 1);
        Debug.Log('Published TypeScript world started: revision 2');
        this.StartCoroutine(this.Blink());
    }

    Update(): void {
        this.elapsed += Time.deltaTime;
        this.transform.Rotate(0, this.speed * Time.deltaTime, 0);
        this.transform.position = new Vector3(0, sampleHeight(this.elapsed), 0);
    }

    private *Blink(): Generator<WaitForSeconds, void, unknown> {
        while (true) {
            this.beacon.SetActive(false);
            yield new WaitForSeconds(this.blinkSeconds);
            this.beacon.SetActive(true);
            yield new WaitForSeconds(this.blinkSeconds);
        }
    }

}

File: KimchilyWebGL/KimchilyCreator/Assets/World/Character.lua
-- Change this file, test with Play, and publish a new QR. No APK rebuild is needed.
local speed = 100
local elapsed = 0

function on_start()
    log("Published Lua world started: revision 2")
    start(function()
        while true do
            refs.beacon.set_active(false)
            wait_seconds(0.3)
            refs.beacon.set_active(true)
            wait_seconds(0.3)
        end
    end)
end

function on_update(dt)
    elapsed = elapsed + dt
    self.rotate(0, speed * dt, 0)
    self.set_position(0, math.sin(elapsed * 2) * 0.12, 0)
end
